tests/run-make/extern-fn-reachable/rmake.rs

```diff
@@ -0,0 +1,45 @@
+//! Smoke test to check that that symbols of `extern "C"` functions and `#[no_mangle]` rust
+//! functions:
+//!
+//! 1. Are externally visible in the dylib produced.
+//! 2. That the symbol visibility is orthogonal to the Rust nameres visibility of the functions
+//!    involved.
+
+//@ ignore-cross-compile
+
+use run_make_support::object::{self, Object};
+use run_make_support::{dynamic_lib_name, is_darwin, path, rfs, rustc};
+
+fn main() {
+    let dylib = dynamic_lib_name("dylib");
+    rustc().input("dylib.rs").output(&dylib).arg("-Cprefer-dynamic").run();
+
+    let expected_symbols = if is_darwin() {
+        // Mach-O states that all exported symbols should have an underscore as prefix. At the
+        // same time dlsym will implicitly add it, so outside of compilers, linkers and people
+        // writing assembly, nobody needs to be aware of this.
+        ["_fun1", "_fun2", "_fun3", "_fun4", "_fun5", "_fun6"]
+    } else {
+        ["fun1", "fun2", "fun3", "fun4", "fun5", "fun6"]
+    };
+
+    let mut found_symbols = vec![];
```

Comment:
why not HashSet that puppy, to get a better log below?

Reply:
Used a `BTreeSet` for ordering, a missing expected symbol would now look like

```
--- stdout -------------------------------
expected_symbols = {"fun1", "fun2", "fun3", "fun4", "fun5", "fun6", "fun7"}
found_symbols = {"fun1", "fun2", "fun3", "fun4", "fun5", "fun6", "rust_metadata_dylib_466e03c52fa389c6"}
------------------------------------------
--- stderr -------------------------------
missing symbol: fun7

thread 'main' panicked at /home/joe/repos/rust/tests/run-make/extern-fn-reachable/rmake.rs:44:9:
missing expected symbols
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------
```